scripts/create_release_tag.sh

```diff
@@ -112,16 +112,93 @@ function update_pre {
     echo "$major.$minor.$patch-pre-$branch.$commits"
 }
 
+function generate_changelog {
+    prev_tag=$1
+    bazel_target=$2
+
+    bug_changelog=''
+    feature_changelog=''
+    cleanup_changelog=''
+
+    # Find all the commits between now and the last release.
+    commits=$(git log HEAD..."$prev_tag" --pretty=format:"%H")
+
+    # Find all file dependencies of the bazel target.
+    bazel query 'kind("source file", deps('"$bazel_target"'))' | sed  -e 's/:/\//' -e 's/^\/\+//' > target_files.txt
+    trap "rm -f target_files.txt" EXIT
+
+    # For each commit, pull out changelog descriptions if it touches any bazel target deps.
+    for commit in $commits
+    do
+        files=$(git show --name-only --format=oneline "$commit" | tail -n +2)
+        for file in $files
+        do
+            if grep -iq "$file" target_files.txt; then
+                changeType=''
+                releaseNote=''
+
+                log=$(git log --format=%B -n 1 "$commit")
+
+                # Get the type of change (cleanup|bug|feature).
+                typeRe='Type of change: /kind ([A-Za-z]+)'
+                if [[ $log =~ $typeRe ]]
+                then
+                changeType=${BASH_REMATCH[1]}
+                fi
+
+                # Get release notes.
+                # shellcheck disable=SC2016
+                # The backticks make shellcheck think that the string should be evaluatable.
+                notesRe='```release-note\s((\S|\s)*)```'
+                if [[ $log =~ $notesRe ]]
+                then
+                releaseNote=${BASH_REMATCH[1]}
+                fi
+
+                if [[ -n $releaseNote ]]
+                then
+                case $changeType in
+                "cleanup")
+                # shellcheck disable=SC2086
+                # Double-quoting adds an extra newline.
+                cleanup_changelog="${cleanup_changelog}- $(echo -e $releaseNote)\n"
+                ;;
+                "bug")
+                # shellcheck disable=SC2086
+                bug_changelog="${bug_changelog}- $(echo -e $releaseNote)\n"
+                ;;
+                "feature")
+                # shellcheck disable=SC2086
+                feature_changelog="${feature_changelog}- $(echo -e $releaseNote)\n"
+                ;;
```

Comment:
IMO this is a bit hard to grok. maybe we should just use arrays and then combine them later?

`feature_changelog=()` to initialize
`feature_changelog+=($releaseNote)` to append
`printf '- %s\n' "${feature_changelog[@]}"` to format+join.

Reply:
I initially tried to do that, but iterating through the array to format the final output got a little tricky. You can set the IFS to `\n` but you can have releaseNotes that are multiline. I guess I could add some special char at the end of each release note and use that as the separator?